### DISCUSSION TITLE
Zephyr Littlefs support for Atmel EEPROM for Non-Volatile Storage

### DISCUSSION THREAD

**Discussion by bob2oneil** [zephyrproject-rtos/zephyr#46939] [Q&A]
We are developing a custom board, and need persistent settings stored and retrieved from non-volatile media using the **littlefs** for file i/o and an EEPROM for the physical media.

While we have some primitives in our driver to read and write to EEPROM, initialize and register - I am unable to build a test application using the following requisite Zephyr kernel configurations which brings in the Littlefs, flash, and flash mapping.

```
CONFIG_FLASH=y
CONFIG_FLASH_MAP=y
CONFIG_FLASH_PAGE_LAYOUT=y
CONFIG_FILE_SYSTEM=y
CONFIG_FILE_SYSTEM_LITTLEFS=y
CONFIG_FS_LOG_LEVEL_DBG=y
CONFIG_FLASH_SHELL=y
```

It could certainly be that our driver is not complete relative to Zephyr integration, and that is what I am hoping to gain some insight into.

Also note that there seems to be a distinction in Zephyr between "flash" and "eeprom" that I do not fully understand all the nuances.  For example, do I need to define an alias for "flash" in the device tree to get the **littlefs** to sit atop this NV storage medium?

The chosen section of the DTS is as follows (does not include a **flash** or **flash_controller** directive).

    chosen {
		zephyr,console = &uart0;
	    zephyr,shell-uart = &uart0;
		zephyr,sram = &sram0;
	};

Our device tree for the EEPROM in terms of **aliases** is as follows:

```
aliases {
        led0 = &green_led;
        led1 = &blue_led;
        io0 = &ext_io0;
        io1 = &ext_io1;
        io2 = &ext_io2;
        io3 = &ext_io3;
        plicirq = &irq;
        eeprom0 = &eeprom_0;
	};
```
	
Where eeprom is as accessed via an I2C interface as follows:

```
&i2c0 {
	status = "okay";
    // clock-frequency = <I2C_BITRATE_FAST>;
    // wp managed by hardware on PE1, no pinctrl, no wp-gpios
	eeprom_0: microchip_24aa@A7 {
		compatible = "microchip,24aa";
		size = <16384>;  //bytes or bits?  Using bytes=16384
        label = "EEPROM";
        reg = <0xAE>;	
		pagesize = <64>;
		address-width = <16>;
		timeout = <10>;
        status = "okay";
	};
```

Our driver is based on the Atmel mc24aa I2C and Atmel AT25 SPI EEPROMs as source file **eeprom_mc24aa.c**, which includes the following device tree definition/registration:

```
// INST_DT_MC24AA_FOREACH(EEPROM_MC24AA_DEVICE);
static const struct eeprom_mc24aa_config eeprom_mc24aa_config = {
    .bus = EEPROM_MC24AA_BUS,
    .size = DT_PROP(INST_DT_MC24AA, size),
    .pagesize = DT_PROP(INST_DT_MC24AA, pagesize), 
    .addr_width = DT_PROP(INST_DT_MC24AA, address_width),
    .readonly = DT_PROP(INST_DT_MC24AA, read_only),
    .timeout = DT_PROP(INST_DT_MC24AA, timeout),
    .bus_is_ready = eeprom_mc24aa_bus_is_ready, 
    .read_fn = eeprom_mc24aa128_read,
    .write_fn = eeprom_mc24aa128_write, 
};

DEVICE_DT_INST_DEFINE(0, &eeprom_mc24aa_init, NULL, NULL,
		    &eeprom_mc24aa_config, POST_KERNEL,
		    CONFIG_EEPROM_INIT_PRIORITY, &eeprom_mc24aa_api);
```

**Comment by henrikbrixandersen** [comment]
Zephyr already contains DTS bindings for generic AT24 (https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/mtd/atmel%2Cat24.yaml) and generic AT25 (https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/mtd/atmel%2Cat25.yaml) EEPROMS - along with a common driver for the two (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/eeprom/eeprom_at2x.c). You can access the EEPROM devices using the Zephyr EEPROM API (https://docs.zephyrproject.org/latest/hardware/peripherals/eeprom.html).

LittleFS has to my knowledge not been ported to use the Zephyr EEPROM API as backend.

**Comment by bob2oneil** [comment]
Thanks for the reply.  Since flash is an offspring of EEPROM, what would you see as the level of effort to achieve a littlefs port?

Do you see this as a small, medium, or large task?

What general directions could you provide.
